pkg/onit/app.go

```diff
@@ -0,0 +1,221 @@
+// Copyright 2019-present Open Networking Foundation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package onit
+
+import (
+	"encoding/json"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// GetSApps returns a list of apps deployed in the cluster
+func (c *ClusterController) GetApps() ([]string, error) {
+	pods, err := c.kubeclient.CoreV1().Pods(c.clusterID).List(metav1.ListOptions{
+		LabelSelector: "type=app",
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	apps := make([]string, len(pods.Items))
+	for i, pod := range pods.Items {
+		apps[i] = pod.Name
+	}
+	return apps, nil
+}
+
+// setupApp creates an app
+func (c *ClusterController) setupApp(name string, config *AppConfig) error {
+	if err := c.createAppConfigMap(name, config); err != nil {
+		return err
+	}
+	if err := c.createAppPod(name); err != nil {
+		return err
+	}
+	if err := c.createAppService(name); err != nil {
+		return err
+	}
+	if err := c.awaitAppReady(name); err != nil {
+		return err
+	}
+	return nil
+}
+
+// createAppConfigMap creates an app configuration
+func (c *ClusterController) createAppConfigMap(name string, config *AppConfig) error {
+	configObj, err := config.load()
+	if err != nil {
+		return err
+	}
+	configJSON, err := json.Marshal(configObj)
+	if err != nil {
+		return err
+	}
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: c.clusterID,
+		},
+		Data: map[string]string{
+			"config.json": string(configJSON),
+		},
+	}
+	_, err = c.kubeclient.CoreV1().ConfigMaps(c.clusterID).Create(cm)
+	return err
+}
+
+// createAppPod creates an app pod
+func (c *ClusterController) createAppPod(name string) error {
+
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: c.clusterID,
+			Labels: map[string]string{
+				"type": "app",
```

Comment:
I am not sure if these labels are good or not. For other subsystems type has a different meaning (e.g. topo, config). Perhaps we should come up with something to be consistent with each other conceptually for both apps and subsystems.

Reply:
I think it makes sense that the `type` is an `app`, doesn't it?

Reply:
I am not sure because "type"  refers to a specific subsystem (topo, config) now and "app" is used as a label ("app", "onos"). If you think it is OK, lets go for it.

Reply:
You have to have the label so that the queries work (see GetApps() for example.